Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+
+        public static ValidationState IsValidDynamic(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            return IsValidDynamic(Instance,
+                                  ref fingerprint,
+                                  ref message,
+                                  ref options);
+        }
+
+        protected override ValidationState IsValidStaticHelper(ref string matchedPattern,
+                                                               ref Dictionary<string, string> groups,
+                                                               ref string failureLevel,
+                                                               ref string message,
+                                                               out Fingerprint fingerprint)
+        {
+            fingerprint = default;
+
+            if (!groups.TryGetNonEmptyValue("secret", out string secret) ||
+                !groups.TryGetNonEmptyValue("id", out string id) ||
+                !groups.TryGetNonEmptyValue("host", out string host))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            if (StaticStringHelper.LikelyPowershellVariable(secret))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            fingerprint = new Fingerprint
+            {
+                Host = host, // technically, this is XML which could contain multiple hosts.
+                Id = id,
+                Secret = secret,
+                Platform = nameof(AssetPlatform.NuGet),
+            };
+
+            return ValidationState.Unknown;
+        }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            string hostXmlAsString = fingerprint.Host;
+            string username = fingerprint.Id;
+            string password = fingerprint.Secret;
+
+            XmlDocument hostXml = new XmlDocument();
+            hostXml.LoadXml(hostXmlAsString);
+
+            // hostXml looks like "<packageSources>...</packageSources>
+            List<string> hosts = hostXml
+                                    ?.ChildNodes[0] // <packageSources>...
+                                        ?.ChildNodes.Cast<XmlNode>().Where(x => x.Name.Equals("add", System.StringComparison.OrdinalIgnoreCase)) // <add ... >  <clear/> (the first node)
+                                            .Select(x => x.Attributes["value"]?.Value ?? x.Attributes["Value"]?.Value).ToList(); // <add key="name of host" value="http://nugetfeedUrl.com" /> (we're looking for URL)
+
+            // Uusually there will probably only be a single host.
+            // In any case, attempt to pass the credentials to each host one by one.
+            using (HttpClient httpClientNoCredentials = CreateHttpClient())
+            using (HttpClient httpClientWithCredentials = CreateHttpClient())
```

Comment:
>httpClientWithCredentials [](start = 30, length = 25)

can you please examine all assignments to a var of this type from CreateHttpClient, here in in out other repo, to establish the consistent name we prefer for this pattern?

our goal is for all rules to look the same, as if authored by a single person, and to be as consistent as possible in style, names, structure. the reason we do this is for readability, but also to lower costs in the future when we decide on a change and attempt to apply mechanism transformations/other parsing of the code.

again, i'm sufficiently old to play the long game here. i don't like going back and making repeated passes over code that has been created organically, with diverse styles, at the point that we realize it's all getting tricky/inconsistent to maintain or update.

Reply:
Merged the two instances into one, renamed simply to "client" as in other validators.

---
In reply to: [611239235](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611239235) [](ancestors = 611239235)